build_modules/lib/src/module_cache.dart

```diff
@@ -53,9 +59,30 @@ class DecodingCache<T> {
   /// content dependencies will be tracked through [reader].
   Future<T> find(AssetId id, AssetReader reader) async {
     if (!await reader.canRead(id)) return null;
-    var result = _cached.putIfAbsent(
-        id, () => Result.capture(reader.readAsBytes(id).then(_fromBytes)));
-    return Result.release(result);
+    _Entry<T> entry;
+    if (!_cached.containsKey(id)) {
+      entry = _cached[id] = _Entry()
+        ..needsCheck = false
+        ..value = Result.capture(reader.readAsBytes(id).then(_fromBytes))
+        ..digest = Result.capture(reader.digest(id));
+    } else {
+      entry = _cached[id];
+      if (entry.needsCheck) {
+        entry.onGoingCheck ??= () async {
+          var previousDigest = await Result.release(entry.digest);
+          entry.digest = Result.capture(reader.digest(id));
+          if (await Result.release(entry.digest) != previousDigest) {
+            entry.value =
+                Result.capture(reader.readAsBytes(id).then(_fromBytes));
+          }
+          entry
+            ..needsCheck = false
+            ..onGoingCheck = null;
+        }();
+        await entry.onGoingCheck;
```

Comment:
[nit]: You could move this await up to the assignment

Reply:
Done